core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -814,10 +814,10 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
 
             if (txnMetadata.producerId != producerId && !retryOnOverflow)
               Left(Errors.INVALID_PRODUCER_ID_MAPPING)
-            else if (!isValidEpoch)
-              Left(Errors.PRODUCER_FENCED)
             else if (txnMetadata.pendingTransitionInProgress && txnMetadata.pendingState.get != PrepareEpochFence)
               Left(Errors.CONCURRENT_TRANSACTIONS)
+            else if (!isValidEpoch)
+              Left(Errors.PRODUCER_FENCED)
```

Comment:
please confirm my understanding:

the mentioned commit/abort + empty abort scenario currently results in fencing the empty abort because the previous commit/abort is left in progress. This change returns concurrent_transactions when performing the empty abort instead. This enables the producer to retry instead of hard failing. Is this correct?

Reply:
exactly! We will retry until the previous operation completes and then we can handle the abort. 👍